cpp/src/arrow/python/arrow_to_python.cc

```diff
@@ -0,0 +1,172 @@
+  auto data = std::dynamic_pointer_cast<UnionArray>(array);                 \
+  int32_t size = array->length();                                           \
+  PyObject* result = CREATE(stop_idx - start_idx);                          \
+  auto types = std::make_shared<Int8Array>(size, data->type_ids());         \
+  auto offsets = std::make_shared<Int32Array>(size, data->value_offsets()); \
+  for (int32_t i = start_idx; i < stop_idx; ++i) {                          \
+    if (data->IsNull(i)) {                                                  \
+      Py_INCREF(Py_None);                                                   \
+      SET_ITEM(result, i - start_idx, Py_None);                             \
+    } else {                                                                \
+      int32_t offset = offsets->Value(i);                                   \
+      int8_t type = types->Value(i);                                        \
+      std::shared_ptr<Array> arr = data->child(type);                       \
+      PyObject* value;                                                      \
+      RETURN_NOT_OK(get_value(arr, offset, type, base, tensors, &value));   \
+      SET_ITEM(result, i - start_idx, value);                               \
+    }                                                                       \
+  }                                                                         \
+  *out = result;                                                            \
+  return Status::OK();
+
+Status DeserializeList(std::shared_ptr<Array> array, int32_t start_idx, int32_t stop_idx,
+    PyObject* base, const std::vector<std::shared_ptr<Tensor>>& tensors, PyObject** out) {
+  DESERIALIZE_SEQUENCE(PyList_New, PyList_SetItem)
+}
+
+Status DeserializeTuple(std::shared_ptr<Array> array, int32_t start_idx, int32_t stop_idx,
+    PyObject* base, const std::vector<std::shared_ptr<Tensor>>& tensors, PyObject** out) {
+  DESERIALIZE_SEQUENCE(PyTuple_New, PyTuple_SetItem)
+}
+
+Status DeserializeDict(std::shared_ptr<Array> array, int32_t start_idx, int32_t stop_idx,
+    PyObject* base, const std::vector<std::shared_ptr<Tensor>>& tensors, PyObject** out) {
+  auto data = std::dynamic_pointer_cast<StructArray>(array);
+  // TODO(pcm): error handling, get rid of the temporary copy of the list
+  PyObject *keys, *vals;
+  PyObject* result = PyDict_New();
+  ARROW_RETURN_NOT_OK(
+      DeserializeList(data->field(0), start_idx, stop_idx, base, tensors, &keys));
+  ARROW_RETURN_NOT_OK(
+      DeserializeList(data->field(1), start_idx, stop_idx, base, tensors, &vals));
+  for (int32_t i = start_idx; i < stop_idx; ++i) {
+    PyDict_SetItem(
+        result, PyList_GetItem(keys, i - start_idx), PyList_GetItem(vals, i - start_idx));
+  }
+  Py_XDECREF(keys);  // PyList_GetItem(keys, ...) incremented the reference count
+  Py_XDECREF(vals);  // PyList_GetItem(vals, ...) incremented the reference count
```

Comment:
So the comment was confusing and I updated it to be more precise. I think the code was correct, please let me know if it makes sense to you with the updated comment -- you are right that keys and vals from DeserializeList should have refcount 1, but then PyDict_SetItem increases the refcount, so we need to manually decrease it again (this is in contrast to PyList_SetItem and PyTuple, which steals the reference and doesn't increase the refcount).